media/201903-report-chalmers.md

```diff
@@ -0,0 +1,750 @@
+configuration.
+
+```k
+    configuration <k>     $PGM:Program </k>
+                  <env>   .Map         </env>
+                  <store> .Map         </store>
+```
+
+-   `<k>` will contain the initial parsed program
+-   `<env>` contains bindings of variable names to store locations
+-   `<store>` contains bindings of store locations to integers
+
+
+Note that the parsed program gets put into the `<k>` cell. Thus, the program is
+part of the execution state, and can be reduced or extended as part of rewrites.
+
+## \K Specifications: Transition Rules ##
+
+The actual operational semantics of a \K program is written as rewrite rules. For example, we might define the following semantic rules for our imperative language.
+
+A transition rule takes the form
+
+```k
+<cellA> X => Y </cellA>
+<cellB> Z => W </cellB>
+etc.
+```
+
+where the arrow represents a rewrite: the value to the left is rewritten to the
+value on the right. The above rule says that "if `<cellA>` contains `X` and `<cellB>` contains `Z`, change the contents of `<cellA>` to `Y` and the contents of `<cellB>` to `W`."
+
+### Syntactic sugar for matching parts of a cell ###
+
+Idiomatically, the `<k>` cell contains an associative
+sequencing operator, `~>`, so the `<k>` cell may look something like
+
+```k
+<k> STMT1 ~> STMT2 ~> STMT3 ~> MORE_STATEMENTS </k>
+```
+
+A rewrite in the `<k>` cell which only cares about the first of these statements
+and replaces it by `NEW` (which could be anything) may then look something like the following:
+
+```k
+<k> (STMT1 ~> THE_OTHER_STATEMENTS)
+    => (NEW ~> THE_OTHER_STATEMENTS)
+</k>
+```
+
+Equivalently, we can apply the rewrite in place:
+
+```k
+<k> (STMT1 => NEW) ~> THE_OTHER_STATEMENTS </k>
+```
+
+To avoid the trouble of explicitly mentioning the entire contents of the cell,
+there is syntacit sugar that lets us write a rule like this is with ellipses,
+which mean "and something else after `~>`".
+
+```k
+<k> STMT1 => NEW ... </k>
+```
+
+Finally, when the configuration contains a Map, we may look up a key-value pair with the key `X` using the following syntax:
+
+```k
+<cell> ... X |-> SX ... </env>
+```
+
+### Variable lookup ###
+
+When an identifier is the first element in the `<k>` cell, rewrite it to its
+corresponding value, by first looking up what store location it is bound to.
+
+```k
+    rule <k> X:Id => V ... </k>
+         <env>   ...  X |-> SX ... </env>
+         <store> ... SX |-> V  ... </store>
+```
+
+### Variable assignment ###
+
+Similarly, variable assigment is done by looking up the store location a
+variable points to and replacing what is there by the right-hand side of the
+assignment.
+
+```k
+    rule <k> X := I:Int => . ... </k>
+         <env>   ...  X |-> SX       ... </env>
+         <store> ... SX |-> (V => I) ... </store>
+```
+
+Note that, for this rule to apply, the right-hand side must be an integer literal, i.e., a fully evaluated expression.
```

Comment:
You may need more explanation of this builtin map sort. Maybe even consider doing as you did above for the associative operator, by expanding it out to show that this actually is:

```
<env> MAP_HEAD   X |-> SX   MAP_TAIL </env>
```

and that the `...` is builtin K syntax which can be used on both sides of an `assoc comm` operator (or just on the tail side of an `assoc` operator).

Reply:
Can it be used for any `assoc` or `assoc comm` operator? I tried playing around with it, and couldn't get it to work if I defined my own operator.

Reply:
Done

Reply:
No, only the builtins unfortunately. We would like more general `...`, but it hasn't been high priority since rarely do people need anything other than `List` (monoid), `Set`, or `Map`.